icefall/decode.py

```diff
@@ -889,3 +893,206 @@ def rescore_with_attention_decoder(
+    tokens = k2.RaggedTensor(tokens_shape, nbest.fsa.tokens)
+    tokens = tokens.remove_values_leq(0)
+
+    #  import pdb
+    #
+    #  pdb.set_trace()
+    path_per_utt = (
+        nbest.shape.row_splits(1)[1:] - nbest.shape.row_splits(1)[:-1]
+    )
+    logging.info(f"path per utt: {path_per_utt}")
+
+    device = model.device
+    device2 = masked_lm_model.device
+
+    alignment = compute_alignment(tokens.to(device2), nbest.shape.to(device2))
+    tgt_ll_list = []
+    for label_name in ["ref_labels", "hyp_labels"]:
+        (
+            masked_src_symbols,
+            src_symbols,
+            tgt_symbols,
+            src_key_padding_mask,
+            tgt_weights,
+        ) = prepare_conformer_lm_inputs(
+            alignment,
+            bos_id=sos_id,
+            eos_id=eos_id,
+            blank_id=blank_id,
+            src_label_name=label_name,
+            unmasked_weight=0.0,
+        )
+
+        masked_src_symbols = masked_src_symbols.to(torch.int64)
+        src_symbols = src_symbols.to(torch.int64)
+        tgt_symbols = tgt_symbols.to(torch.int64)
+
+        masked_lm_memory, masked_lm_pos_emb = masked_lm_model(
+            masked_src_symbols, src_key_padding_mask
+        )
+
+        tgt_nll = masked_lm_model.decoder_nll(
+            masked_lm_memory,
+            masked_lm_pos_emb,
+            src_symbols,
+            tgt_symbols,
+            src_key_padding_mask,
+        )
+
+        # nll means negative log-likelihood
+        # ll means log-likelihood
+        tgt_ll = -1 * (tgt_nll * tgt_weights).sum(dim=-1)
+
+        tgt_ll_list.append(tgt_ll)
+
+    # hyp - ref
+    masked_lm_scores = tgt_ll_list[1] - tgt_ll_list[0]
+    masked_lm_scores = masked_lm_scores.to(device)
+
+    # TODO(fangjun): Support passing a ragged tensor to `decoder_nll` directly.
+    token_ids = tokens.tolist()
+
+    nll = model.decoder_nll(
+        memory=expanded_memory,
+        memory_key_padding_mask=expanded_memory_key_padding_mask,
+        token_ids=token_ids,
+        sos_id=sos_id,
+        eos_id=eos_id,
+    )
+    assert nll.ndim == 2
+    assert nll.shape[0] == len(token_ids)
+
+    attention_scores = -nll.sum(dim=1)
+
+    if ngram_lm_scale is None:
+        ngram_lm_scale_list = [0.01, 0.05, 0.08]
+        ngram_lm_scale_list += [0.1, 0.3, 0.5, 0.6, 0.7, 0.9, 1.0]
+        ngram_lm_scale_list += [1.1, 1.2, 1.3, 1.5, 1.7, 1.9, 2.0]
+        ngram_lm_scale_list += [2.1, 2.2, 2.3, 2.5, 3.0, 4.0, 5.0]
+    else:
+        ngram_lm_scale_list = [ngram_lm_scale]
+
+    if attention_scale is None:
+        attention_scale_list = [0.01, 0.05, 0.08]
+        attention_scale_list += [0.1, 0.3, 0.5, 0.6, 0.7, 0.9, 1.0]
+        attention_scale_list += [1.1, 1.2, 1.3, 1.5, 1.7, 1.9, 2.0]
+        attention_scale_list += [2.1, 2.2, 2.3, 2.5, 3.0, 4.0, 5.0]
+    else:
+        attention_scale_list = [attention_scale]
+
+    if masked_lm_scale is None:
+        masked_lm_scale_list = [0.01, 0.05, 0.08]
```

Comment:
Perhaps, now that there are so many scales, we should start pruning the scales to the more frequently seen values.

Reply:
Yes, I agree. It takes hours to compute the final WERs with so many scales.